goclone

```diff
@@ -76,39 +81,50 @@ done
 exeargs=$*
 
 # is github_user/github_repo
-if [[ "$src" =~ ^([A-Za-z0-9-]+)\/([A-Za-z0-9-]+)$ ]]; then
+if [[ "$src" =~ ^([A-Za-z0-9_-]+)\/([A-Za-z0-9_-]+)(\@([A-Za-z0-9\._\-]+))?$ ]]; then
     user="${BASH_REMATCH[1]}"
     repo="${BASH_REMATCH[2]}"
+    branch_name="${BASH_REMATCH[4]}"
+    [[ ! -z "${BASH_REMATCH[4]}" ]] && src="${src:0:$(( ${#src} - ${#BASH_REMATCH[3]} ))}"
     github="true"
 
 # is git@github.com:user/repo.git
-elif [[ "$src" =~ ^git@github\.com:([A-Za-z0-9-]+)\/([A-Za-z0-9-]+)\.git$ ]]; then
+elif [[ "$src" =~ ^git@github\.com:([A-Za-z0-9_-]+)\/([A-Za-z0-9_-]+)\.git(\@([A-Za-z0-9\._\-]+))?$ ]]; then
     user="${BASH_REMATCH[1]}"
     repo="${BASH_REMATCH[2]}"
+    branch_name="${BASH_REMATCH[4]}"
+    [[ ! -z "${BASH_REMATCH[4]}" ]] && src="${src:0:$(( ${#src} - ${#BASH_REMATCH[3]} ))}"
```

Comment:
I don't follow what's happening here at all. Could it be worth commenting?

Reply:
1. `[[ ! -z "${BASH_REMATCH[4]}" ]]` only trim `$src` if we actually pulled out a branch name.

2. `${src:0:$(( ${#src} - ${#BASH_REMATCH[3]} ))}` is bash trim syntax: `parameter:start:end`

3. `$(( ${#src} - ${#BASH_REMATCH[3]} ))` is `len(src) - len(bash_rematch[3])` (`bash_rematch[3]` is the branch name with the `@` still intact)